sql/core/src/main/scala/org/apache/spark/sql/ml/FrequentItems.scala

```diff
@@ -0,0 +1,124 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package org.apache.spark.sql.ml
+
+
+import org.apache.spark.sql.catalyst.plans.logical.LocalRelation
+import org.apache.spark.sql.types.{StructType, ArrayType, StructField}
+
+import scala.collection.mutable.{Map => MutableMap}
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.{Row, DataFrame, functions}
+
+private[sql] object FrequentItems extends Logging {
+
+  /**
+   * Merge two maps of counts. Subtracts the sum of `otherMap` from `baseMap`, and fills in
+   * any emptied slots with the most frequent of `otherMap`.
+   * @param baseMap The map containing the global counts
+   * @param otherMap The map containing the counts for that partition
+   * @param maxSize The maximum number of counts to keep in memory
+   */
+  private def mergeCounts[A](
```

Comment:
I think the implementation could be cleaner if we wrap `MutableMap[A, Long]` with a utility class:

~~~scala
class FreqItemCounter(size: k) {
  def add(any: Any, count: Long = 1L): this.type
  def merge(other: FreqItemCounter): this.type = {
    other.toSeq.foreach { case (k, c) =>
      add(k, c)
    }
  }
  def items: Array[Any]
  def toSeq: Seq[(Any, Long)]
}
~~~